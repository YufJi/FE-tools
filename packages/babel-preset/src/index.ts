import type { 
  ConfigAPI,
  SimpleCacheKey,
} from '@babel/core';

import type { Options as EnvOptions  } from '@babel/preset-env';

type ReactOptions = {
  runtime?: 'automatic' | 'classic',
  development?: boolean,
  importSource?: string,
  useBuiltIns?: boolean,
  throwIfNamespace?: boolean,
  useSpread?: boolean,
  pragma?: string,
  pragmaFrag?: string,
}

interface Options {
  cache?: () => SimpleCacheKey,
  env?: EnvOptions,
  react?: ReactOptions,
}

export function preset(api: ConfigAPI, options: Options = {}) {
  api.cache.using(options.cache || (() => false));

  const {
    env,
    react,
  } = options;

  const presets = [
    [require.resolve('@babel/preset-env'), env],
    [require.resolve('@babel/preset-react'), react]
  ];

  const plugins = [
    // 下面两个的顺序的配置都不能动
    [require.resolve('@babel/plugin-proposal-decorators'), { legacy: true }],
    [require.resolve('@babel/plugin-proposal-class-properties'), { loose: true }],

    require.resolve('@babel/plugin-proposal-export-default-from'),
    [require.resolve('@babel/plugin-proposal-pipeline-operator'), { proposal: 'minimal' }],
    require.resolve('@babel/plugin-proposal-do-expressions'),
    require.resolve('@babel/plugin-proposal-function-bind'),
  ];

  return {
    presets,
    plugins,
  };
}

export default preset;
