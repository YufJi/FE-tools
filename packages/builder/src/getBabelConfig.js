/* eslint-disable func-names */

export default function (opts) {
  const {
    target, typescript, type, runtimeHelpers, filePath, browserFiles, nodeFiles, nodeVersion,
  } = opts;

  let isBrowser = target === 'browser';
  // rollup 场景下不会传入 filePath
  if (filePath) {
    if (isBrowser) {
      if (nodeFiles.includes(filePath)) isBrowser = false;
    } else if (browserFiles.includes(filePath)) isBrowser = true;
  }

  const targets = isBrowser ? {
    chrome: '51',
    ie: '9',
  } : { node: nodeVersion || 6 };

  return {
    presets: [
      ...(typescript ? [require.resolve('@babel/preset-typescript')] : []),
      [require.resolve('@babel/preset-env'), {
        loose: true,
        targets,
        modules: type === 'esm' ? false : (type === 'cjs' && !isBrowser) ? 'cjs' : 'auto',
      }],
      ...(isBrowser ? [require.resolve('@babel/preset-react')] : []),
    ],
    plugins: [
      require.resolve('@babel/plugin-proposal-export-default-from'),
      require.resolve('@babel/plugin-proposal-do-expressions'),
      [require.resolve('@babel/plugin-proposal-decorators'), { legacy: true }],
      [require.resolve('@babel/plugin-proposal-class-properties'), { loose: true }],
      ...(runtimeHelpers
        ? [[require.resolve('@babel/plugin-transform-runtime'), { ...runtimeHelpers, useESModules: isBrowser && (type === 'esm') }]]
        : []),
    ],
  };
}
