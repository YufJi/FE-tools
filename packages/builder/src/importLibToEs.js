import { join, dirname } from 'path';
import fs from 'fs';
import signale from 'signale';

const cwd = process.cwd();

function replacePath(path) {
  if (path.node.source && /\/lib\//.test(path.node.source.value)) {
    const esModule = path.node.source.value.replace('/lib/', '/es/');
    const esPath = dirname(join(cwd, `node_modules/${esModule}`));
    if (fs.existsSync(esPath)) {
      signale.info(`[es build] replace ${path.node.source.value} with ${esModule}`);
      path.node.source.value = esModule;
    }
  }
}

export default function replaceLib() {
  return {
    visitor: {
      ImportDeclaration: replacePath,
      ExportNamedDeclaration: replacePath,
    },
  };
}
