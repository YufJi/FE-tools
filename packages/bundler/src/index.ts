#!/usr/bin/env node

import * as path from 'path';
import { Command } from 'commander';
import { dev } from './dev';
import { build } from './build';

export * from './types';

const program = new Command();

program
  .name('bundler')
  .description('Bundler for JavaScript and TypeScript')
  .usage('[command] [options]');

program
  .command('dev')
  .description('Start the devServer')
  .option('--root <root>', 'Root directory')
  .option('-c, --config <config>', 'Path to the config file', 'bundler.config.js')
  .action((options) => {
    const cwd = process.cwd();
    const root = path.resolve(cwd, options.root) || cwd;
    
    dev({
      root,
      config: options.config,
    });
  });

program
  .command('build')
  .description('Build the project')
  .option('--root <root>', 'Root directory')
  .option('-c, --config <config>', 'Path to the config file', 'bundler.config.js')
  .action((options) => {
    const cwd = process.cwd();
    const root = path.resolve(cwd, options.root) || cwd;

    build({
      root,
      config: options.config,
    });
  });

program.parse(process.argv);
