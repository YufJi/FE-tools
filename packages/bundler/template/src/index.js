import React from 'react';
import ReactDOM from 'react-dom';

import './index.less';

class App extends React.Component {
  state = {
    name: 'hello curiosity'
  }

  render() {
    const { name } = this.state;
    return (
      <div className="name">{name}</div>
    )
  }
}


ReactDOM.render(<App />, document.getElementById('root'))
