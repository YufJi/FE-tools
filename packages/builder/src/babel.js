/* eslint-disable func-names */
import { join, extname, relative } from 'path';
import { existsSync, readFileSync, statSync } from 'fs';
import vfs from 'vinyl-fs';
import signale from 'signale';
import rimraf from 'rimraf';
import through from 'through2';
import slash from 'slash';
import chokidar from 'chokidar';
import babel from '@babel/core';
import gulpTs from 'gulp-typescript';
import gulpLess from 'gulp-less';
import gulpIf from 'gulp-if';
import getBabelConfig from './getBabelConfig';

export default async function (opts) {
  const {
    cwd,
    rootPath,
    type,
    watch,
    importLibToEs,
    bundleOpts,
  } = opts;
  const {
    target = 'browser',
    runtimeHelpers,
    extraBabelPresets = [],
    extraBabelPlugins = [],
    browserFiles = [],
    nodeFiles = [],
    nodeVersion,
    disableTypeCheck,
    lessInBabelMode,
  } = bundleOpts;

  const srcPath = join(cwd, 'src');
  const targetDir = type === 'esm' ? 'es' : 'lib';
  const targetPath = join(cwd, targetDir);

  signale.info(`Clean ${targetDir} directory`);
  rimraf.sync(targetPath);

  function transform(opts) {
    const { file, type } = opts;
    signale.info(`[${type}] Transform: ${slash(file.path).replace(`${cwd}/`, '')}`);

    const babelOpts = getBabelConfig({
      target,
      type,
      typescript: true,
      runtimeHelpers,
      filePath: slash(relative(cwd, file.path)),
      browserFiles,
      nodeFiles,
      nodeVersion,
    });
    if (importLibToEs && type === 'esm') {
      babelOpts.plugins.push(require.resolve('./importLibToEs'));
    }
    babelOpts.presets.push(...extraBabelPresets);
    babelOpts.plugins.push(...extraBabelPlugins);

    return babel.transform(file.contents, {
      ...babelOpts,
      filename: file.path,
    }).code;
  }

  function getTSConfig() {
    const tsconfigPath = join(cwd, 'tsconfig.json');
    const templateTsconfigPath = join(__dirname, '../template/tsconfig.json');

    if (existsSync(tsconfigPath)) {
      return JSON.parse(readFileSync(tsconfigPath, 'utf-8')).compilerOptions || {};
    }
    if (rootPath && existsSync(join(rootPath, 'tsconfig.json'))) {
      return JSON.parse(readFileSync(join(rootPath, 'tsconfig.json'), 'utf-8')).compilerOptions || {};
    }
    return JSON.parse(readFileSync(templateTsconfigPath, 'utf-8')).compilerOptions || {};
  }

  function createStream(src) {
    const tsConfig = getTSConfig();
    const babelTransformRegexp = disableTypeCheck ? /\.(t|j)sx?$/ : /\.jsx?$/;

    function isTsFile(path) {
      return /\.tsx?$/.test(path) && !path.endsWith('.d.ts');
    }

    function isTransform(path) {
      return babelTransformRegexp.test(path) && !path.endsWith('.d.ts');
    }

    return vfs
      .src(src, {
        allowEmpty: true,
        base: srcPath,
      })
      // 开启类型校验且是ts文件
      .pipe(gulpIf((f) => !disableTypeCheck && isTsFile(f.path), gulpTs(tsConfig)))
      // 处理less文件
      .pipe(gulpIf((f) => lessInBabelMode && /\.less$/.test(f.path), gulpLess({ ...lessInBabelMode })))
      .pipe(gulpIf(
        (f) => isTransform(f.path),
        through.obj((file, env, cb) => {
          try {
            file.contents = Buffer.from(transform({
              file,
              type,
            }));
            // .jsx -> .js
            file.path = file.path.replace(extname(file.path), '.js');
            cb(null, file);
          } catch (e) {
            signale.error(`Compiled faild: ${file.path}`);
            console.log(e);
            cb(null);
          }
        }),
      ))
      .pipe(vfs.dest(targetPath));
  }

  return new Promise((resolve) => {
    createStream([
      join(srcPath, '**/*'),
    ]).on('end', () => {
      if (watch) {
        signale.info('Start watch', srcPath);
        chokidar
          .watch(srcPath, {
            ignoreInitial: true,
          })
          .on('all', (event, fullPath) => {
            const relPath = fullPath.replace(srcPath, '');

            signale.info(`[${event}] ${join(srcPath, relPath)}`);

            if (!existsSync(fullPath)) return;

            if (statSync(fullPath).isFile()) {
              createStream([fullPath]);
            }
          });
      }
      resolve();
    });
  });
}
