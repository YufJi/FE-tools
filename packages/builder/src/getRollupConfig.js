/* eslint-disable func-names */
import { basename, extname, join } from 'path';
import { terser } from 'rollup-plugin-terser';
import commonjs from '@rollup/plugin-commonjs';
import babel from '@rollup/plugin-babel';
import replace from '@rollup/plugin-replace';
import json from '@rollup/plugin-json';
import nodeResolve from '@rollup/plugin-node-resolve';
import typescript from '@rollup/plugin-typescript';
import postcss from 'rollup-plugin-postcss';
import NpmImport from 'less-plugin-npm-import';
import { camelCase } from 'lodash';
import tempDir from 'temp-dir';
import autoprefixer from 'autoprefixer';
import svgr from '@svgr/rollup';

import getBabelConfig from './getBabelConfig';

export default function (opts) {
  const { type, entry, cwd, importLibToEs, bundleOpts } = opts;
  const {
    umd,
    esm,
    cjs,
    file,
    target = 'browser',
    extractCSS = false,
    injectCSS = true,
    cssModules: modules,
    extraPostCSSPlugins = [],
    extraBabelPresets = [],
    extraBabelPlugins = [],
    autoprefixer: autoprefixerOpts,
    include = /node_modules/,
    runtimeHelpers: runtimeHelpersOpts,
    replace: replaceOpts,
    nodeVersion,
    typescriptOpts,
    nodeResolveOpts = {},
    lessInRollupMode = {},
  } = bundleOpts;
  const entryExt = extname(entry);
  const name = file || basename(entry, entryExt);
  const isTypeScript = /.tsx?/.test(entryExt);
  const extensions = ['.js', '.jsx', '.ts', '.tsx', '.es6', '.es', '.mjs'];

  let pkg = {};
  try {
    pkg = require(join(cwd, 'package.json')); // eslint-disable-line
  } catch (e) {
    console.log(e);
  }

  // cjs 不给浏览器用，所以无需 runtimeHelpers
  const runtimeHelpers = type === 'cjs' ? false : runtimeHelpersOpts;
  const babelOpts = {
    ...getBabelConfig({
      type,
      target: type === 'esm' ? 'browser' : target,
      typescript: false,
      runtimeHelpers,
      nodeVersion,
    }),
    babelHelpers: 'bundled',
    exclude: /\/node_modules\//,
    babelrc: false,
    // ref: https://github.com/rollup/rollup-plugin-babel#usage
    extensions,
  };
  if (importLibToEs && type === 'esm') {
    babelOpts.plugins.push(require.resolve('./importLibToEs'));
  }
  babelOpts.presets.push(...extraBabelPresets);
  babelOpts.plugins.push(...extraBabelPlugins);

  // rollup configs
  const input = join(cwd, entry);
  const format = type;

  // ref: https://rollupjs.org/guide/en#external
  // 潜在问题：引用包的子文件时会报 warning，比如 @babel/runtime/helpers/esm/createClass
  // 解决方案：可以用 function 处理
  const external = [
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.peerDependencies || {}),
  ];
  // umd 只要 external peerDependencies
  const externalPeerDeps = Object.keys(pkg.peerDependencies || {});

  function testExternal(pkgs, id) {
    return pkgs.includes(id);
  }

  const terserOpts = {
    compress: {
      pure_getters: true,
      unsafe: true,
      unsafe_comps: true,
      warnings: false,
    },
  };
  const plugins = [
    svgr(),
    postcss({
      extract: extractCSS,
      inject: injectCSS,
      modules,
      use: [
        [
          'less',
          {
            plugins: [new NpmImport({ prefix: '~' })],
            javascriptEnabled: true,
            ...lessInRollupMode,
          },
        ],
      ],
      plugins: [
        autoprefixer(autoprefixerOpts),
        ...extraPostCSSPlugins,
      ],
    }),
    ...(replaceOpts && Object.keys(replaceOpts || {}).length ? [replace(replaceOpts)] : []),
    nodeResolve({
      mainFields: ['module', 'jsnext:main', 'main'],
      extensions,
      ...nodeResolveOpts,
    }),
    ...(isTypeScript
      ? [
        typescript({
          cacheDir: `${tempDir}/.rollup.tscache`,
          tsconfig: join(cwd, 'tsconfig.json'),
          ...(typescriptOpts || {}),
        }),
      ]
      : []),
    babel(babelOpts),
    json(),
  ];

  switch (type) {
    case 'esm':
      return [
        {
          input,
          output: {
            format,
            file: join(cwd, `dist/${(esm && esm.file) || `${name}.esm`}.js`),
          },
          plugins: [...plugins, ...(esm && esm.minify ? [terser(terserOpts)] : [])],
          external: testExternal.bind(null, external),
        },
        ...(esm && esm.mjs
          ? [
            {
              input,
              output: {
                format,
                file: join(cwd, `dist/${(esm && esm.file) || `${name}`}.mjs`),
              },
              plugins: [
                ...plugins,
                replace({
                  'process.env.NODE_ENV': JSON.stringify('production'),
                }),
                terser(terserOpts),
              ],
              external: testExternal.bind(null, externalPeerDeps),
            },
          ]
          : []),
      ];

    case 'cjs':
      return [
        {
          input,
          output: {
            exports: 'named',
            format,
            file: join(cwd, `dist/${(cjs && cjs.file) || name}.js`),
          },
          plugins: [...plugins, ...(cjs && cjs.minify ? [terser(terserOpts)] : [])],
          external: testExternal.bind(null, external),
        },
      ];

    case 'umd':
      // Add umd related plugins
      plugins.push(commonjs({
        include,
      }));

      return [
        {
          input,
          output: {
            format,
            file: join(cwd, `dist/${(umd && umd.file) || `${name}.umd`}.js`),
            globals: umd && umd.globals,
            name: (umd && umd.name) || (pkg.name && camelCase(basename(pkg.name))),
          },
          plugins: [
            ...plugins,
            replace({
              'process.env.NODE_ENV': JSON.stringify('development'),
            }),
          ],
          external: testExternal.bind(null, externalPeerDeps),
        },
        ...(umd && umd.minFile === true
          ? [
            {
              input,
              output: {
                format,
                file: join(cwd, `dist/${(umd && umd.file) || `${name}.umd`}.min.js`),
                globals: umd && umd.globals,
                name: (umd && umd.name) || (pkg.name && camelCase(basename(pkg.name))),
              },
              plugins: [
                ...plugins,
                replace({
                  'process.env.NODE_ENV': JSON.stringify('production'),
                }),
                terser(terserOpts),
              ],
              external: testExternal.bind(null, externalPeerDeps),
            },
          ] : []),
      ];

    default:
      throw new Error(`Unsupported type ${type}`);
  }
}
