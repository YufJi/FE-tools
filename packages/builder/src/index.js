import { join } from 'path';
import rimraf from 'rimraf';
import { merge } from 'lodash';
import signale from 'signale';
import rollup from './rollup';
import babel from './babel';
import { getExistFile } from './utils';
import getUserConfig from './getUserConfig';

function getBundleOpts(opts) {
  const { cwd } = opts;
  const entry = getExistFile({
    cwd,
    files: ['src/index.tsx', 'src/index.ts', 'src/index.jsx', 'src/index.js'],
    returnRelative: true,
  });
  const userConfig = getUserConfig({ cwd });
  const userConfigs = Array.isArray(userConfig) ? userConfig : [userConfig];
  return userConfigs.map((userConfig) => {
    const bundleOpts = merge(
      {
        entry,
      },
      userConfig,
    );

    // Support config esm: 'rollup' and cjs: 'rollup'
    if (typeof bundleOpts.esm === 'string') {
      bundleOpts.esm = { type: bundleOpts.esm };
    }
    if (typeof bundleOpts.cjs === 'string') {
      bundleOpts.cjs = { type: bundleOpts.cjs };
    }

    return bundleOpts;
  });
}

async function build(opts) {
  const { cwd, watch, rootPath } = opts;
  // Get user config
  const bundleOptsArray = getBundleOpts(opts);
  for (const bundleOpts of bundleOptsArray) {
    // Clean dist
    signale.info('Clean dist directory');
    rimraf.sync(join(cwd, 'dist'));

    // Build umd
    if (bundleOpts.umd) {
      signale.info('Build umd');
      await rollup({
        cwd,
        type: 'umd',
        entry: bundleOpts.entry,
        watch,
        bundleOpts,
      });
    }

    // Build cjs
    if (bundleOpts.cjs) {
      const { cjs } = bundleOpts;
      signale.info(`Build cjs with ${cjs.type}`);
      if (cjs.type === 'babel') {
        await babel({ cwd, rootPath, watch, type: 'cjs', bundleOpts });
      } else {
        await rollup({
          cwd,
          type: 'cjs',
          entry: bundleOpts.entry,
          watch,
          bundleOpts,
        });
      }
    }

    // Build esm
    if (bundleOpts.esm) {
      const { esm } = bundleOpts;
      signale.info(`Build esm with ${esm.type}`);
      const importLibToEs = esm && esm.importLibToEs;
      if (esm && esm.type === 'babel') {
        await babel({ cwd, rootPath, watch, type: 'esm', importLibToEs, bundleOpts });
      } else {
        await rollup({
          cwd,
          type: 'esm',
          entry: bundleOpts.entry,
          importLibToEs,
          watch,
          bundleOpts,
        });
      }
    }
  }
}

export default async function (options) {
  await build(options);
}
