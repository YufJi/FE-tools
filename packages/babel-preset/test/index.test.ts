import * as babel from '@babel/core';
import { preset } from '../src/index';

describe('Custom Babel Preset', () => {
  it('transforms ES6 code to ES5 as expected', () => {
    const code = 'const a = 1; let b = 2;';
    const result = babel.transform(code, {
      presets: [
        preset
      ],
    });

    expect(result?.code).toMatchSnapshot();
  });

  it('transforms JSX to React.createElement', () => {
    const jsxCode = '<div>Hello, world!</div>';
    const result = babel.transform(jsxCode, {
      presets: [
        preset
      ],
    });

    expect(result?.code).toMatchSnapshot();
  });
});
