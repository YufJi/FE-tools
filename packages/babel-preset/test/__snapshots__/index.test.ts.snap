// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Babel Preset transforms ES6 code to ES5 as expected 1`] = `
""use strict";

var a = 1;
var b = 2;"
`;

exports[`Custom Babel Preset transforms JSX to React.createElement as expected 1`] = `
""use strict";

/*#__PURE__*/React.createElement("div", null, "Hello, world!");"
`;
